cartographer/mapping_2d/sparse_pose_graph.cc

```diff
@@ -444,23 +438,25 @@ void SparsePoseGraph::RunOptimization() {
   const auto& node_data = optimization_problem_.node_data();
   for (int trajectory_id = 0;
        trajectory_id != static_cast<int>(node_data.size()); ++trajectory_id) {
-    int node_data_index = 0;
     const int num_nodes = trajectory_nodes_.num_indices(trajectory_id);
-    int node_index = optimization_problem_.num_trimmed_nodes(trajectory_id);
-    for (; node_data_index != static_cast<int>(node_data[trajectory_id].size());
-         ++node_data_index, ++node_index) {
-      const mapping::NodeId node_id{trajectory_id, node_index};
-      trajectory_nodes_.at(node_id).pose = transform::Embed3D(
-          node_data[trajectory_id][node_data_index].point_cloud_pose);
+    int last_node_data_index = 0;
+    for (auto it = node_data.at(trajectory_id).begin();
+         it != node_data.at(trajectory_id).end(); ++it) {
+      const mapping::NodeId node_id{trajectory_id, it->first};
+      trajectory_nodes_.at(node_id).pose =
+          transform::Embed3D(it->second.point_cloud_pose);
+      last_node_data_index = it->first;
```

Comment:
Why update this? In the end this is `node_data.at(trajectory_id).rbegin()->first` or `0` if the trajectory contains no optimized nodes.